### PR TITLE
Styleguide programmeobject

### PR DESCRIPTION
**Pull request by tatitati**
# PLEASE DONT REVIEW. WORK IN PROGRESS

Builders for:
- programme items:
      Episode (refactoring)
      Clip
- programme containers:
      Brand
      Series

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/Builders/AbstractCoreEntityBuilder.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+namespace App\Builders;
+
+use BBC\ProgrammesPagesService\Domain\Entity\Image;
+use BBC\ProgrammesPagesService\Domain\Entity\MasterBrand;
+use BBC\ProgrammesPagesService\Domain\Entity\Options;
+use BBC\ProgrammesPagesService\Domain\Entity\Programme;
+use BBC\ProgrammesPagesService\Domain\ValueObject\Pid;
+use BBC\ProgrammesPagesService\Domain\ValueObject\Synopses;
+
+Abstract class AbstractCoreEntityBuilder
```
Comment (votemike) [inline src/Builders/AbstractCoreEntityBuilder.php]: Lower case?

File: src/Builders/BrandBuilder.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
+<?php
+
+namespace App\Builders;
+
+use BBC\ProgrammesPagesService\Domain\Entity\Brand;
+
+class BrandBuilder extends AbstractProgrammeContainerBuilder implements BuilderInterface
+{
+    protected function __construct()
```
Comment (votemike) [inline src/Builders/BrandBuilder.php]: Is this useful?

File: src/Builders/AbstractCoreEntityBuilder.php (hunk 1 above)
Comment (votemike) [inline src/Builders/AbstractCoreEntityBuilder.php]: Should this include a `build()` method?

File: src/Builders/BrandBuilder.php (hunk 2 above)
Comment (tatitati) [reply src/Builders/BrandBuilder.php]: I want to avoid wrong use of the builders. So I'm trying to avoid someone doing something like:
```$clip = new ClipBuilder()```

Instead I want something more friendly like:
```$clip = ClipBuilder::bigClip()->build()```
or 
```$clip = ClipBuilder::default()->build()```

In this way tests become more expressive

File: src/Builders/AbstractCoreEntityBuilder.php (hunk 1 above)
Comment (tatitati) [reply src/Builders/AbstractCoreEntityBuilder.php]: No. AbstractBuilders are "programmeItem", 'ProgrammeContainer'. But we don't want to build those objects. Instead we want to build Clips, episodes, brands, series, etc. Abstract classes are used in here just to reduce configuration of default values in childs. Otherwise these classes would have lot of default values and maintain our tests would be hard.

File: src/Builders/BrandBuilder.php (hunk 2 above)
Comment (votemike) [reply src/Builders/BrandBuilder.php]: Could you add a comment in the code making that clear?

File: src/Builders/AbstractCoreEntityBuilder.php (hunk 1 above)
Comment (votemike) [reply src/Builders/AbstractCoreEntityBuilder.php]: So you don't want `abstract public function build()`?

Comment (tatitati) [reply src/Builders/AbstractCoreEntityBuilder.php]: So far I'm using an interface BuildInterface.

Comment (votemike) [reply src/Builders/AbstractCoreEntityBuilder.php]: I can't see `BuildInterface` anywhere in this PR...

File: src/Builders/BrandBuilder.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
+<?php
+
+namespace App\Builders;
+
+use BBC\ProgrammesPagesService\Domain\Entity\Brand;
+
+class BrandBuilder extends AbstractProgrammeContainerBuilder
+{
+    protected function __construct()
```
Comment (votemike) [inline src/Builders/BrandBuilder.php]: Why is this needed?

File: src/Builders/ClipBuilder.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public static function default()
+    {
+        return new self();
+    }
+
+    public function build()
+    {
+        return new CLip(
```
Comment (votemike) [inline src/Builders/ClipBuilder.php]: `Clip`?

File: src/Builders/SeriesBuilder.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        $this->title = 'My series title';
+        $this->searchTitle = 'My search title';
+    }
+
+    /**
+     * Create a builder with default state.
+     * This default state reduce the amount of steps necessary when trying to get a final entity in an specified state
+     */
+    public static function default()
```
Comment (votemike) [inline src/Builders/SeriesBuilder.php]: Could this be moved to the parent?

Comment (tatitati) [reply src/Builders/SeriesBuilder.php]: I just tried it. The parent is abstract, the app crash saying that is not possible to instantiate an abstract class.

Comment (votemike) [reply src/Builders/SeriesBuilder.php]: Does it work with `new static()` instead of `new self()`?

File: src/Builders/ClipBuilder.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+{
+    protected function __construct()
+    {
+        parent::__construct();
+
+        $this->title = 'my clip title';
+        $this->searchTitle = 'my search clip title';
+    }
+
+    public static function default()
```
Comment (votemike) [inline src/Builders/ClipBuilder.php]: Do you need this method if the parent has it?

File: src/Builders/SeriesBuilder.php (hunk 5 above)
Comment (votemike) [inline src/Builders/SeriesBuilder.php]: Do you need this method if the parent has it?
